### PR TITLE
[geneus_main.py] Resolve package dependencies with attention to the order

### PR DESCRIPTION
**Pull request by wkentaro**
This is to solve these issue::
* https://github.com/jsk-ros-pkg/geneus/pull/16
* https://github.com/start-jsk/2014-semi/issues/196
* https://github.com/jsk-ros-pkg/jsk_roseus/issues/257

### CODE REVIEW

File: src/geneus/geneus_main.py [hunk 1]
```diff
@@ -72,9 +72,9 @@ def package_depends_impl(pkg, depends=[]):
         return depends
 
     tmp_depends = [x for x in rp.get_depends(pkg, False) if x not in depends]
-    depends.extend(tmp_depends)
+    depends = tmp_depends + depends
```
Comment (wkentaro) [inline src/geneus/geneus_main.py]: to load dependencies first and after that load the pkg.

**Comment by k-okada** [comment]
!!! Thanks!!!

File: src/geneus/geneus_main.py [hunk 2]
```diff
@@ -72,9 +72,9 @@ def package_depends_impl(pkg, depends=[]):
...
     for p in tmp_depends:
-        package_depends_impl(p, depends)
+        depends = package_depends_impl(p, depends)
```
Comment (wkentaro) [inline src/geneus/geneus_main.py]: to update depends.

File: src/geneus/geneus_main.py [hunk 3]
```diff
@@ -105,7 +105,7 @@ def genmain(argv, progname):
             pkg = args[1]               # ARG_PKG
             pkg_dependences = package_depends(pkg) + args[2:]   # args[1] ARG_PKG
             # load all dependences and then load target package
-            for p in list(set(pkg_dependences)):
+            for p in pkg_dependences:
```
Comment (wkentaro) [inline src/geneus/geneus_main.py]: `set` breaks the order so don't use.

**Comment by wkentaro** [comment]
This modification may still not solve the problem, when the package has deep implicit dependencies.
So **DON'T MERGE THIS PR** now.

That is difficult to solve dependent order with removing duplicate in function `package_depends_impl`.

So why not to get all dependencies first, and then check the order?
This process solves::
* dependencies (all packages)
* duplicates
* dependent order

I wrote the code::
https://github.com/wkentaro/geneus/commit/3aedcd2d0e80090ec0c0afb4a2c36396b0d571e6

**Comment by k-okada** [comment]
Ok, so can we merge this?

**Comment by wkentaro** [comment]
If the `ros::load-ros-package` fails depending on the order, this new code also can cause same error in some case.
I'd like to close this PR and have you merge #18 instead.

**Comment by wkentaro** [comment]
I close this. See #18

**Pull request closed without merging**